src/notification/mod.rs

```diff
@@ -0,0 +1,399 @@
+use axum::{
+    extract::{Query, State},
+    http::StatusCode,
+    response::IntoResponse,
+    Json,
+};
+use base64::{engine::general_purpose::URL_SAFE, Engine as _};
+use bitflags::bitflags;
+use chrono::{DateTime, Duration, Utc};
+use nostr::nips::nip19::FromBech32;
+use rand::RngCore;
+use serde::{Deserialize, Serialize};
+use tracing::{error, warn};
+
+use crate::{notification::email::build_email_confirmation_message, util, AppState};
+
+pub mod email;
+pub mod notification_store;
+
+/// Maximum age of a challenge - we expect requests to be made immediately after each other
+const CHALLENGE_EXPIRY_SECONDS: i64 = 120; // 2 minutes
+
+/// Maximum age of an email confirmation
+const EMAIL_CONFIRMATION_EXPIRY_SECONDS: i64 = 60 * 60 * 24; // 1 day
+
+/// A challenge to validate the request comes from a given npub
+#[derive(Debug)]
+pub struct Challenge {
+    pub npub: String,
+    pub challenge: String,
+    pub created_at: DateTime<Utc>,
+}
+
+/// Email confirmation state
+#[derive(Debug)]
+pub struct EmailConfirmation {
+    pub npub: String,
+    pub email: String,
+    pub confirmed: bool,
+    pub sent_at: DateTime<Utc>,
+}
+
+/// Email preferences state
+#[derive(Debug)]
+#[allow(unused)]
+pub struct EmailPreferences {
+    pub npub: String,
+    pub enabled: bool,
+    pub email: String,
+    pub email_confirmed: bool,
+    pub ebill_url: url::Url,
+    pub flags: PreferencesFlags,
+}
+
+bitflags! {
+/// A set of preference flags packed in an efficient way
+#[derive(Debug, Clone, PartialEq, Eq)]
+    pub struct PreferencesFlags: i64 {
+        const IssueBill = 0b0001;
+    }
+}
+
+impl Default for PreferencesFlags {
+    fn default() -> Self {
+        Self::IssueBill
+    }
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct NotificationStartReq {
+    pub npub: String,
+}
+
+#[derive(Debug, Clone, Serialize)]
+pub struct NotificationStartResp {
+    pub challenge: String,
+}
```

Comment:
Nitpick, could have the ttl as return field.

Reply:
good point :+1: